pkg/storagev2/leveldbstore/transaction.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package leveldbstore
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+)
+
+var _ storage.TxStore = (*TxStore)(nil)
+
+// TxStore is an implementation of in-memory Store
+// where all Store operations are done in a transaction.
+type TxStore struct {
+	*storage.TxStoreBase
+
+	// Bookkeeping of invasive operations executed
+	// on the Store to support rollback functionality.
+	batchMu sync.Mutex
+	batch   *leveldb.Batch
+}
+
+// Put implements the Store interface.
+func (s *TxStore) Put(item storage.Item) error {
+	err := s.TxStoreBase.Put(item)
```

Comment:
So one problem with this approach is we are updating the underlying store directly. This means we are on a slippery slope as if there is some unexpected problem and we fail to rollback, we are screwed. I was initially trying to see if we could write the updates in a batch, but the problem is if we write it in the batch, we wont be able to read the latest state of the keys that were updated in the batch. And this is required. So I guess we cannot get around this.

With my approach, problem was we will have to create some hybrid read operations (Get, Has... etc) which could read from the batch. But then iteration is not easy to do and it starts getting complicated. And seems like we depend on this, the way we have the components. For eg in pinning we iterate and delete in batches. So the delete has to prevent the iteration from reading the same entries again. And this would be slightly challenging. So if we go with the current approach we dont have all these complications but the risk of going into some invalid state. We should also test if we can handle context cancellations properly. If not we should maybe write the undo batch in some recovery file and replay it once we are up. This comment is mainly to highlight this fact. I have no perfect solution here.